aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -240,32 +260,43 @@ std::tuple<Tensor&, Tensor&> cummax_out(Tensor& values, Tensor& indices, const T
     values.resize_(self.sizes());
     indices.resize_(self.sizes());
     if(self.dim() == 0) {
-      values.fill_(self.item());
+      values.fill_(self);
       indices.fill_(0);
-    }
-    else if(self.numel() != 0) {
-      // update values and indices for the first values along the dimension dim
-      values.narrow(dim, 0, 1) = self.narrow(dim, 0, 1);
-      indices.narrow(dim, 0, 1).fill_(0);
-      for(int i = 1; i < self.size(dim); i++) {
-        auto res_at_i = at::max(at::cat({values.narrow(dim, i-1, 1), self.narrow(dim, i, 1)}, dim), dim, true);
-        // values at index i
-        values.narrow(dim, i, 1) = std::get<0>(res_at_i);
-        // indices at index i
-        indices.narrow(dim, i, 1) = at::max(indices.narrow(dim, i-1, 1), (i * (std::get<1>(res_at_i))));
-       }
+    } else if(self.numel() != 0) {
+      dim = maybe_wrap_dim(dim, self.dim());
+      at::_cummax_helper(self, values, indices, dim);
     }
   }
   namedinference::propagate_names(values, self);
   namedinference::propagate_names(indices, self);
-  return std::tuple<Tensor &,Tensor &>{values, indices};
+  return std::forward_as_tuple(values, indices);
 }
 
 std::tuple<Tensor, Tensor> cummax(const Tensor& self, int64_t dim) {
   auto values = at::empty(self.sizes(), self.options());
   auto indices = at::empty(self.sizes(), self.options().dtype(at::kLong));
   at::cummax_out(values, indices, self, dim);
-  return std::tuple<Tensor &,Tensor &>{values, indices};
+  return std::make_tuple(values, indices);
+}
+
+void cummin_helper_cpu(const Tensor& self, Tensor& values, Tensor& indices, int64_t dim) {
+    AT_DISPATCH_ALL_TYPES_AND2(at::ScalarType::Bool, at::ScalarType::Half,
+      self.scalar_type(), "cummin_cpu", [&]() -> void {
+      at::native::tensor_dim_apply3<scalar_t, scalar_t>(self, values, indices, dim,
+        [](const scalar_t* self_data, scalar_t* values_data, int64_t* indices_data,
+          int self_dim_size, int self_stride, int values_stride, int indices_stride) {
+            auto cummin = self_data[0];
+            int idx = 0;
+            for(int i = 0; i < self_dim_size; i++) {
+              if(self_data[i*self_stride] <= cummin) {
```

Comment:
Same as with cummax, you should correctly handle `nan`, you can take a look at how comparison for ArgMax/ArgMin is implemented in SharedReduceOps.h. Also, please add tests for tensors containing nans.

Reply:
ArgReductionOps expect int64_t index so do you think we should let index be defined as int64_t(As is for max and other ops) or templatize it and define the equivalent logic for ArgMinOps(for cummin) by:
```
if(std::isnan(curr_elem) || (!std::isnan(cummin) && std::less_equal<scalar_t>()(curr_elem, cummin))) {
        cummin = self_data[i*self_stride];
        idx = i;
}
```

Reply:
Templatizing and defining op similar to ArgMin sounds good to me, but I don't have a strong opinion here.